### PR TITLE
Display promotion items in place of the announcements

### PR DESCRIPTION
**Pull request by ClementPasteau**
Stories: 
Slideshow https://gdevelop-storybook.s3.amazonaws.com/announcement-mobile-friendly/latest/index.html?path=/story/ui-building-blocks-slideshow--loading
Promotions https://gdevelop-storybook.s3.amazonaws.com/announcement-mobile-friendly/latest/index.html?path=/story/promotionsslideshow--default

image
image

### CODE REVIEW

File: newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js [hunk 1]
```diff
@@ -2,21 +2,31 @@
...
 import {
   type Announcement,
+  type Promotion,
   listAllAnnouncements,
+  listAllPromotions,
 } from '../Utils/GDevelopServices/Announcement';
 import { ANNOUNCEMENTS_FETCH_TIMEOUT } from '../Utils/GlobalFetchTimeouts';
 
+// NOTE: For the moment, we store announcements and promotions in the same context,
+// as previously both were mixed under the announcenments endpoint.
```
Comment (4ian) [inline newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js:12]: typo ;)

File: newIDE/app/src/fixtures/GDevelopServicesTestData/index.js [hunk 2]
```diff
@@ -2783,3 +2786,36 @@ export const fakeAssetPackLicenses = [
...
+    imageUrl:
+      'https://resources.gdevelop.io/announcements/GDevelops_Mega_Pack_Updated.png',
+    mobileImageUrl:
+      'https://resources.gdevelop.io/announcements/GDevelops_Mega_Pack_Mobile.jpg',
+    productId: '43994a30-c54b-4f5d-baf5-6e1f99b13824',
+    display: 'all',
+    type: 'asset-pack',
+  },
+  {
+    id: 'premium-featuring-bubble-dogs', // TO REMOVE ON 26/02/24
```
Comment (4ian) [inline newIDE/app/src/fixtures/GDevelopServicesTestData/index.js:2802]: Maybe avoid having something in test data that is was not made by GDevelop?

File: newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js [hunk 3]
```diff
@@ -31,35 +41,43 @@ export const AnnouncementsFeedStateProvider = ({
...
-        } catch (error) {
-          console.error(
-            `Unable to load the announcements from the api:`,
-            error
-          );
-          setError(error);
-        }
+        // Logic to remove once promotions are displayed to enough users.
+        // For now, we filter out promotions from the announcements.
+        const filteredAnnouncements = fetchedAnnouncements.filter(
```
Comment (AlexandreSi) [inline newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js:62]: Should you add the check we do a bit everywhere "Array.isArray"?

**Review by 4ian — APPROVED** [review]
Quick review but overall lgtm

File: newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js (hunk 3 above)
Comment (4ian) [reply newIDE/app/src/AnnouncementsFeed/AnnouncementsFeedContext.js:62]: Probably should be in the function doing the actual API call

File: newIDE/app/src/MainFrame/EditorContainers/HomePage/BuildSection/index.js [hunk 4]
```diff
@@ -427,25 +427,27 @@ const BuildSection = ({
...
-                <BackgroundText>
-                  <Trans>
-                    Create your first project using one of our templates or
-                    start from scratch.
-                  </Trans>
-                </BackgroundText>
-              </Paper>
-            </Column>
-          </ListItem>
+          <Line>
```
Comment (AlexandreSi) [inline newIDE/app/src/MainFrame/EditorContainers/HomePage/BuildSection/index.js:430]: Why this change? A display problem? Do the added margins look ok?

File: newIDE/app/src/UI/Slideshow/Slideshow.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+import * as React from 'react';
+import { Column, Line, marginsSize } from '../../UI/Grid';
+import Paper from '../../UI/Paper';
+import Skeleton from '@material-ui/lab/Skeleton';
+import {
+  useResponsiveWindowWidth,
+  type WidthType,
+} from '../../UI/Reponsive/ResponsiveWindowMeasurer';
+import Measure from 'react-measure';
+import { createStyles, makeStyles } from '@material-ui/core';
```
Comment (AlexandreSi) [inline newIDE/app/src/UI/Slideshow/Slideshow.js:11]:
```suggestion
import { createStyles, makeStyles } from '@material-ui/core/styles';
```

File: newIDE/app/src/MainFrame/EditorContainers/HomePage/BuildSection/index.js (hunk 4 above)
Comment (ClementPasteau) [reply newIDE/app/src/MainFrame/EditorContainers/HomePage/BuildSection/index.js:430]: Yes, on mobile & desktop the empty message is too close to the buttons:
image

File: newIDE/app/src/UI/Slideshow/Slideshow.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+      if (!items || items.length === 1) return;
+
+      setCurrentSlide(currentSlide === items.length - 1 ? 0 : currentSlide + 1);
+    },
+    [currentSlide, items]
+  );
+
+  React.useEffect(
+    () => {
+      const interval = setInterval(() => {
```
Comment (AlexandreSi) [inline newIDE/app/src/UI/Slideshow/Slideshow.js:102]: You could use `useInterval` instead

File: newIDE/app/src/UI/Slideshow/Slideshow.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    [handleRightArrowClick]
+  );
+
+  if (!items) {
+    // If they're loading, display a skeleton so that it doesn't jump when loaded.
+    return (
+      <Measure
+        bounds
+        onResize={contentRect => {
+          setContainerWidth(contentRect.bounds.width);
```
Comment (AlexandreSi) [inline newIDE/app/src/UI/Slideshow/Slideshow.js:116]: Is it normal that it's not the same onResize when there is an item displayed?

File: newIDE/app/src/UI/Slideshow/Slideshow.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        }}
+      >
+        {({ contentRect, measureRef }) => (
+          <Paper square background="dark">
+            <div
+              ref={measureRef}
+              style={{
+                display: 'flex',
+                flex: 1,
```
Comment (AlexandreSi) [inline newIDE/app/src/UI/Slideshow/Slideshow.js:125]: Should you put most of the styles in the canonical `styles` object?

File: newIDE/app/src/stories/componentStories/AnnouncementsFeed/AnnouncementsFeed.stories.js [hunk 9]
```diff
@@ -48,8 +53,9 @@ export const ErrorLoadingAnnouncements = () => {
     <AnnouncementsFeedContext.Provider
       value={{
         announcements: null,
+        promotions: null,
         error: new Error('Fake error'),
-        fetchAnnouncements: action('fetchAnnouncements'),
+        fetchAnnouncementsAndPromotions: action('fetchAnnouncements'),
```
Comment (AlexandreSi) [inline newIDE/app/src/stories/componentStories/AnnouncementsFeed/AnnouncementsFeed.stories.js:58]:
```suggestion
        fetchAnnouncementsAndPromotions: action('fetchAnnouncementsAndPromotions'),
```
And in the other places also

**Review by AlexandreSi — APPROVED** [review]
LGTM

File: newIDE/app/src/UI/Slideshow/Slideshow.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    [currentSlide, items]
+  );
+
+  const handleRightArrowClick = React.useCallback(
+    () => {
+      if (!items || items.length === 1) return;
+
+      setCurrentSlide(currentSlide === items.length - 1 ? 0 : currentSlide + 1);
+    },
+    [currentSlide, items]
```
[the remark this answers is not in the document]
Comment (ClementPasteau) [reply newIDE/app/src/UI/Slideshow/Slideshow.js:97]: I actually didn't know this was possible 😨  nice

File: newIDE/app/src/UI/Slideshow/Slideshow.js (hunk 7 above)
Comment (ClementPasteau) [reply newIDE/app/src/UI/Slideshow/Slideshow.js:116]: Yes the fact that the image is positioned absolutely makes the size of the container different when it's a skeleton or the image (the margins are either not counted or are)
It's a bit confusing, but ensure how to fix that

File: newIDE/app/src/UI/Slideshow/Slideshow.js (hunk 8 above)
Comment (ClementPasteau) [reply newIDE/app/src/UI/Slideshow/Slideshow.js:125]: oh yes, forgot to refactor this

File: newIDE/app/src/fixtures/GDevelopServicesTestData/index.js (hunk 2 above)
Comment (ClementPasteau) [reply newIDE/app/src/fixtures/GDevelopServicesTestData/index.js:2802]: yep changing this
